### PR TITLE
Add documentation to the module

### PR DESCRIPTION
**Pull request by danielgraycode**
I've added some documentation to make it easier for users to understand and use the module.

### CODE REVIEW

**Comment by FozzieHi** [comment]
@simontabor Why isn't this merged?

File: lib/2FA.js [hunk 1]
```diff
@@ -5,6 +5,16 @@ var QR = require('qr-image');
 var TFA = module.exports = {};
 
 // get a base36 crypto secure key
+    /**
+     * @description
+     * Generates the key which can be used to make a QR code for the user to add to their authenticator apps and check codes against.
+     * @param length The length desired for the key.
+     * @returns A key you can store, check codes and generate QR codes from.
```
Comment (simontabor) [inline lib/2FA.js]: This function doesn't return anything, it accepts another parameter called `cb`

File: lib/2FA.js [hunk 2]
```diff
@@ -5,6 +5,16 @@ var QR = require('qr-image');
 var TFA = module.exports = {};
 
 // get a base36 crypto secure key
```
Comment (simontabor) [inline lib/2FA.js]: may as well remove the old comments now we have lovely JSDoc :)

File: lib/2FA.js [hunk 3]
```diff
@@ -25,6 +35,22 @@ TFA.generateKey = function(length, cb) {
   get();
 };
 
+    /**
+     * @description
+     * Verify a users HOTP code they entered against the given key
+     * @param key  The key to check the HOTP code against
+     * @param code The code entered by the user from the authenticator app.
+     * @param counter The time for verifying the code against the key.
```
Comment (simontabor) [inline lib/2FA.js]: missing `opts` param for drift options, if you'd like i can add this

File: lib/2FA.js [hunk 4]
```diff
@@ -25,6 +35,22 @@ TFA.generateKey = function(length, cb) {
...
+     * @returns Either true or false depending on if the code is valid or not.
+     * @example
+     * let counter = Math.floor(Date.now() / 1000 / 30);
```
Comment (simontabor) [inline lib/2FA.js]: this project is ES5 so probably best to document using `var` for now

File: lib/2FA.js [hunk 5]
```diff
@@ -25,6 +35,22 @@ TFA.generateKey = function(length, cb) {
...
+     * let validCode = tfa.veryifyHOTP(key, code, counter);
```
Comment (simontabor) [inline lib/2FA.js]:
```suggestion
     * var validCode = tfa.verifyHOTP(key, code, counter);
```

File: lib/2FA.js [hunk 6]
```diff
@@ -25,6 +35,22 @@ TFA.generateKey = function(length, cb) {
...
+     * if (validCode == true) {
+     * //Log user in
+     * } else {
+     * //Display erorr message
```
Comment (simontabor) [inline lib/2FA.js]:
```suggestion
     * // Display error message
```

File: lib/2FA.js [hunk 7]
```diff
@@ -94,6 +134,15 @@ TFA.generateCode = function(key, counter, opts) {
...
 };
 
+    /**
+     * @description
+     * Converts a key to a format which can be pasted into an authenticator app.
+     * @param key  The key to convert
+     * @returns The authenticater-friendly code.
+     * @example
+     * let multikey = tfa.base32Encode(key);
+     * console.log("You can manually enter this code into the authenticator app: " & multikey)
```
Comment (simontabor) [inline lib/2FA.js]:
```suggestion
     * console.log('You can manually enter this code into the authenticator app: ' + multikey)
```

File: lib/2FA.js [hunk 8]
```diff
@@ -43,6 +69,20 @@ TFA.verifyHOTP = function(key, code, counter, opts) {
...
+     * Verify a users TOTP code they entered against the given key
+     * @param key  The key to check the TOTP code against
+     * @param code The code entered by the user from the authenticator app.
+     * @returns Either true or false depending on if the code is valid or not.
+     * @example
+     * let validCode = tfa.verifyTOTP(key, code);
+     * if (validCode == true) {
+     * //Log user in
+     * } else {
+     * //Display erorr message
```
Comment (simontabor) [inline lib/2FA.js]:
```suggestion
     * // Display error message
```

**Comment by danielgraycode** [comment]
@simontabor Any update on this? It's been a while and I noticed it still hasn't merged.

File: lib/2FA.js [hunk 9]
```diff
@@ -104,6 +153,18 @@ TFA.generateUrl = function (name, account, key) {
            + '&secret=' + TFA.base32Encode(key);
 };
 
+    /**
+     * @description
+     * Generates a QR code that can be scanned to add the account to an authenciator app.
+     * @param name The name to be listed on the authenticator app- usually the company or product name.
+     * @param account The account to be listed on the authenticator app- Usually the users email.
+     * @param key The key generated for the users multifactor.
```
Comment (simontabor) [inline lib/2FA.js]: missing optional opts parameter here

File: lib/2FA.js [hunk 10]
```diff
@@ -136,6 +197,17 @@ TFA.generateGoogleQR = function(name, account, key, opts, cb) {
   });
 };
 
+    /**
+     * @description
+     * Generates codes which can be checked against and used if the authenticator app is not available.
+     * @param count The number of backup keys you want to generate
+     * @param pattern The pattern you want the codes to be generated in
+     * @returns An array of backup codes you can store and show the user
```
Comment (simontabor) [inline lib/2FA.js]: calls back, not returns

File: lib/2FA.js [hunk 11]
```diff
@@ -136,6 +197,17 @@ TFA.generateGoogleQR = function(name, account, key, opts, cb) {
   });
 };
 
+    /**
+     * @description
+     * Generates codes which can be checked against and used if the authenticator app is not available.
+     * @param count The number of backup keys you want to generate
+     * @param pattern The pattern you want the codes to be generated in
```
Comment (simontabor) [inline lib/2FA.js]: Pattern is optional, defaults to `'xxxx-xxxx-xxxx'`

**Review by simontabor — CHANGES_REQUESTED** [review]
Hi Daniel, thanks for the contribution! Sorry for the incredibly slow review - I'd done the review then forgot to submit it. My bad entirely.

There are a few things to be fixed up before we merge. Thanks!

**Comment by danielgraycode** [comment]
Hi Simon!
Thanks for the suggestions, I’ll work on getting this done and will soon push another commit.

**Comment by danielgraycode** [comment]
I've changed things according to your feedback, and pushed the commit with these changes.

**Comment by danielgraycode** [comment]
Any update on this Simon?

**Comment by danielgraycode** [comment]
??

**Review by simontabor — APPROVED** [review]

**Comment by simontabor** [comment]
Merged! @danielgraycode sorry for taking so long to do this, really appreciate the contribution!

**Pull request merged**